src/dochandle.rs

```diff
@@ -87,9 +98,28 @@ impl DocHandle {
         F: FnOnce(&mut Automerge) -> T,
     {
         let res = {
-            let mut state = self.shared_document.write();
-            f(&mut state.automerge)
+            let (start_heads, new_heads, res) = {
+                let mut state = self.shared_document.write();
+                let start_heads = state.automerge.get_heads();
+                let res = f(&mut state.automerge);
+                let new_heads = state.automerge.get_heads();
+                (start_heads, new_heads, res)
+            };
+
+            let doc_changed = start_heads != new_heads;
+
+            // Always note the last heads seen by the handle,
+            // for use with `changed`.
+            *self.last_heads.lock() = new_heads;
```

Comment:
Does this not need to be inside the block above to ensure that no concurrent process interleaves updating the last heads here?

Reply:
So actually each handle has it's own mutex, it's not actually shared between clones of the same handle. I only put it in a mutex otherwise the `with_doc_{}` methods would have required a mutable reference to the handle. I propose we review the API later and perhaps change it to require a mutable reference.